runners/jet/src/main/java/org/apache/beam/runners/jet/DAGBuilder.java

```diff
@@ -0,0 +1,223 @@
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package org.apache.beam.runners.jet;
+
+import com.hazelcast.jet.core.DAG;
+import com.hazelcast.jet.core.Edge;
+import com.hazelcast.jet.core.Processor;
+import com.hazelcast.jet.core.ProcessorMetaSupplier;
+import com.hazelcast.jet.core.Vertex;
+import com.hazelcast.jet.function.FunctionEx;
+import com.hazelcast.jet.function.SupplierEx;
+import java.io.ByteArrayInputStream;
+import java.util.ArrayList;
+import java.util.Collection;
+import java.util.Collections;
+import java.util.HashMap;
+import java.util.HashSet;
+import java.util.List;
+import java.util.Map;
+import java.util.Objects;
+import org.apache.beam.sdk.coders.Coder;
+import org.apache.beam.sdk.util.WindowedValue;
+import org.apache.beam.sdk.values.KV;
+
+/** Utility class for wiring up Jet DAGs based on Beam pipelines. */
+public class DAGBuilder {
+
+  private final DAG dag = new DAG();
+  private final int localParallelism;
+
+  private final Map<String, Vertex> edgeStartPoints = new HashMap<>();
+  private final Map<String, List<Vertex>> edgeEndPoints = new HashMap<>();
+  private final Map<String, Coder> edgeCoders = new HashMap<>();
+  private final Map<String, String> pCollsOfEdges = new HashMap<>();
+
+  private final List<WiringListener> listeners = new ArrayList<>();
+
+  private int vertexId = 0;
+
+  DAGBuilder(JetPipelineOptions options) {
+    this.localParallelism = options.getJetLocalParallelism();
+  }
+
+  DAG getDag() {
+    wireUp();
+    return dag;
+  }
+
+  void registerConstructionListeners(WiringListener listener) {
+    listeners.add(listener);
+  }
+
+  String newVertexId(String transformName) {
+    return vertexId++ + " (" + transformName + ")";
+  }
+
+  void registerCollectionOfEdge(String edgeId, String pCollId) {
+    String prevPCollId = pCollsOfEdges.put(edgeId, pCollId);
+    if (prevPCollId != null) {
+      throw new RuntimeException("Oops!");
+    }
+  }
+
+  void registerEdgeStartPoint(String edgeId, Vertex vertex, Coder coder) {
+    Objects.requireNonNull(edgeId);
+    Objects.requireNonNull(vertex);
+    Objects.requireNonNull(coder);
+
+    Vertex prevVertex = edgeStartPoints.put(edgeId, vertex);
+    if (prevVertex != null) {
+      throw new RuntimeException("Oops!");
+    }
+
+    Coder prevCoder = edgeCoders.put(edgeId, coder);
+    if (prevCoder != null) {
+      throw new RuntimeException("Oops!");
+    }
+  }
+
+  void registerEdgeEndPoint(String edgeId, Vertex vertex) {
+    edgeEndPoints.computeIfAbsent(edgeId, x -> new ArrayList<>()).add(vertex);
+  }
+
+  Vertex addVertex(String id, ProcessorMetaSupplier processorMetaSupplier) {
+    return dag.newVertex(id, processorMetaSupplier);
+  }
+
+  Vertex addVertex(String id, SupplierEx<Processor> processor) {
+    return dag.newVertex(id, processor)
+        .localParallelism(
+            localParallelism) // todo: quick and dirty hack for now, can't leave it like this
```

Comment:
Why is this a hack?

Reply:
It's just a left-over TODO item, removing it in the next commit.